server/plugins/OscUGens.cpp

```diff
@@ -533,12 +533,34 @@ void TWindex_Ctor(TWindex* unit) {
     } else {
         SETCALC(TWindex_next_k);
     }
-    unit->m_prevIndex = 0;
-    unit->m_trig = -1.f; // make it trigger the first time
-    TWindex_next_k(unit, 1);
+    // Generate a random value here for the initialization & first sample
```

Comment:
would it perhaps be easier to make sure that the condition `trig > 0.f && unit->m_trig <= 0.f)` is fulfilled, then call `TWindex_next_k` and then set `trig = 0` again? Then you don't have to duplicate all the code. Usually it is done like this in other cases.

Reply:
> would it perhaps be easier to make sure that the condition trig > 0.f && unit->m_trig <= 0.f) is fulfilled

We can control the condition `unit->m_trig <= 0.f`, but not ` trig > 0.f` because `trig` is read from the input. The solution I have here is required, from what I can tell, by the case where there is not a trigger on the first input frame.

So I get around this by at least making this test _false_ on the first frame by use of `unit->m_trig = 1.f`, so I can control the value of both the initialization and first sample (so a new random number isn't generated). I agree duplicating code isn't pretty :-/

> Usually it is done like this in other cases.
The unique thing here is that randomness is used in the calc function, whereas many other UGens don't involve randomness.

I'll put a bit more thought into it just in case...

Reply:
yes, it would be great to have a solution that works without so much code duplication. If there is none, well then that is also a good result – but we may have to check other UGens that need randomness in the calc function. How is it with `TRand`?

Reply:
`TRand` also sets the init sample directly in the Ctor without calling a `next_` function:
https://github.com/supercollider/supercollider/blob/0fe26b63d3c10add37fbc270789535aab3737b1a/server/plugins/NoiseUGens.cpp#L512-L514
In that case there isn't much overhead in terms of using that randomly generated number (just scaling it between `hi` and `lo`)—it _is_ the output value. (At a quick glance this UGen will also need to be updated to fix the init sample bug.)

Whereas most of the duplicated code is how the generated number is used: normalizing and scaling it then using it in the loop that chooses the index to output.

An alternative would be to make a helper function that does all of this, so it could be factored out of the Ctor and both calc functions. But this would be a departure from UGen style generally, and I'm not sure what kind of overhead that function might add to the calc function.

Having a look though _NoiseUGens.cpp_, many of the UGens that use `mRGen` do so in the Ctor and set the output sample directly.

Reply:
Here's a version using a helper function `TWindex_chooseNewIndex`:
```
static int32 TWindex_chooseNewIndex(TWindex* unit) {
    int maxindex = unit->mNumInputs;
    int32 index = maxindex;
    float normalize = ZIN0(1);
    float maxSum = unit->m_maxSum;
    
    if (maxSum < 0.f) {
        maxSum = 0.f;
        if (normalize == 1) {
            for (int32 k = 2; k < maxindex; ++k) {
                maxSum += ZIN0(k);
            }
        } else
            maxSum = 1.f;
        unit->m_maxSum = maxSum;
    }
    RGen& rgen = *unit->mParent->mRGen;
    
    float max = maxSum * rgen.frand();
    float sum = 0.f;
    for (int32 k = 2; k < maxindex; ++k) {
        sum += ZIN0(k);
        if (sum >= max) {
            index = k - 2;
            break;
        }
    }
    return index;
}

void TWindex_Ctor(TWindex* unit) {
    if (INRATE(0) == calc_FullRate) {
        SETCALC(TWindex_next_a);
    } else {
        SETCALC(TWindex_next_k);
    }
    unit->m_maxSum = -1.f; // trigger update of maxSum
    int32 index = TWindex_chooseNewIndex(unit);
    OUT0(0) = index;
    unit->m_prevIndex = index;
    // Ensure a first-sample trigger doesn't cause another rand val
    // so initialization sample will equal first output sample
    unit->m_trig = 1.f;
}

void TWindex_next_k(TWindex* unit, int inNumSamples) {
    float trig = ZIN0(0);
    float* out = ZOUT(0);
    int32 index;
    unit->m_maxSum = -1.f;
    if (trig > 0.f && unit->m_trig <= 0.f) {
        index = TWindex_chooseNewIndex(unit);
        unit->m_prevIndex = index;
    } else {
        index = unit->m_prevIndex;
    }
    LOOP1(inNumSamples, ZXP(out) = index;)
    unit->m_trig = trig;
}

void TWindex_next_a(TWindex* unit, int inNumSamples) {
    float* trig = ZIN(0);
    float* out = ZOUT(0);
    int32 index;
    float curtrig;
    unit->m_maxSum = -1.f;

    LOOP1(
        inNumSamples, curtrig = ZXP(trig); if (curtrig > 0.f && unit->m_trig <= 0.f) {
            index = TWindex_chooseNewIndex(unit);
            unit->m_prevIndex = index;
        } else index = unit->m_prevIndex;

        ZXP(out) = index; unit->m_trig = curtrig;)
}
```
It's a more substantial overhaul. I could push this change for discussion in the Review, though if we revert, it may involve a force-push, and I'm not sure what that does to the Review history. I could also revert without force-pushing, and the commits could be squashed on merge.

If you could let me know the etiquette for force-push in this situation and that's the preferred way, I'll remove this post and we could continue the review that way.

Thanks for following up with this, @telephon!